topi/include/topi/reduction.h

```diff
@@ -0,0 +1,387 @@
+ *
+ * \param data The input tensor.
+ * \param axis The axes along which the reduction is performed.
+ * \param func The reduction function eg. tvm::sum
+ * \param keepdims If this is set to true, the axes which are reduced are
+ * left in the result as dimensions with size one. This enables the result
+ * to broadcast correctly against the input array.
+ *
+ * \return The result tensor.
+ */
+Tensor CommReduce(const Tensor& data,
+                  const std::vector<int>& axis,
+                  FReduce func,
+                  bool keepdims = false) {
+  auto ndim = data->shape.size();
+  CHECK_NE(ndim, 0) << "Cannot reduce a 0 dim Tensor";
+  auto real_axis = GetRealAxis(static_cast<int>(ndim), axis);
+
+  Array<IterVar> reduce_axes;
+  for (auto i : real_axis) {
+    std::string name = "k" + std::to_string(i);
+    reduce_axes.push_back(
+      tvm::reduce_axis(Range(0, data->shape[i]), name));
+  }
+
+  Array<Expr> target_shape;
+  if (keepdims) {
+    for (int i = 0; i < ndim; ++i) {
+      if (std::find(real_axis.begin(), real_axis.end(), i) != real_axis.end()) {
+        // real_axis contains i
+        target_shape.push_back(1);
+      } else {
+        target_shape.push_back(data->shape[i]);
+      }
+    }
+  } else {
+    for (int i = 0; i < ndim; ++i) {
+      if (std::find(real_axis.begin(), real_axis.end(), i) == real_axis.end()) {
+        // real_axis does not contain i
+        target_shape.push_back(data->shape[i]);
+      }
+    }
+  }
+
+  auto compute = [ndim, keepdims, &real_axis, &reduce_axes, &func, &data]
+  (const Array<Var>& indices) {
+    Array<Expr> eval_range;
+    Array<Var> eval_indices;
+    int arg_counter = 0;
+    int red_counter = 0;
+
+    for (int i = 0; i < ndim; ++i) {
+      if (std::find(real_axis.begin(), real_axis.end(), i) != real_axis.end()) {
+        // real_axis contains i
+        eval_range.push_back(reduce_axes[red_counter]);
+        eval_indices.push_back(reduce_axes[red_counter]->var);
+        red_counter++;
+      } else {
+        if (!keepdims) {
+          eval_range.push_back(indices[arg_counter]);
+          arg_counter++;
+        } else {
+          eval_range.push_back(indices[i]);
+        }
+      }
+    }
+
+    return func(data(eval_range), reduce_axes);
+  };
+
+  return tvm::compute(target_shape, compute, data->op->name + "_red", kCommReduce);
+}
+
+/*!
+* \brief Create an index reduction operation.
+*
+* \param data The input tensor.
+* \param axis The axes along which the reduction is performed.
+* \param func The reduction function
+* \param keepdims If this is set to true, the axes which are reduced are
+* left in the result as dimensions with size one. This enables the result
+* to broadcast correctly against the input array.
+*
+* \return The result tensor.
+*/
+Tensor CommReduceIdx(const Tensor& data,
+                     const std::vector<int>& axis,
+                     FCommReduce func,
+                     bool keepdims = false) {
```

Comment:
Is it possible to merge `CommReduceIdx` and `CommReduce`? Most of the codes are the same.

Reply:
They are very similar, but annoyingly different in some control flow. An approach I was thinking about was pulling shared code out into separate functions - eg. functions to calculate reduce_axes and target_shape. That would cut down much of the duplication. Would that be ok?

Reply:
Yes, that should be okay.

Reply:
any update wrt this comment?

Reply:
I've pulled most of the duplication out into 2 common functions: MakeReduceAxes and MakeReduceTargetShape, they should be in the latest commit.